packages/@react-stately/numberfield/src/useNumberFieldState.ts

```diff
@@ -105,9 +104,17 @@ export function useNumberFieldState(
   // Update the input value when the number value or format options change. This is done
   // in a useEffect so that the controlled behavior is correct and we only update the
   // textfield after prop changes.
-  useLayoutEffect(() => {
+  let [prevValue, setPrevValue] = useState(numberValue);
+  let [prevLocale, setPrevLocale] = useState(locale);
+  let [prevFormatOptions, setPrevFormatOptions] = useState(formatOptions);
```

Comment:
These should be refs, not state. Setting them should not trigger a re-render. You could potentially also combine them into a single ref, that's an object containing them all?

Reply:
if they haven't changed, then setting then won't cause a re-render
we only get in here if one of them has changed, and we already call setState on the main one we're interested in
the setStates will get aggregated and result in only one new render

Note, this is the way that the react docs say to implement it as well https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops

Reply:
but I still don't get why it needs to be a state and not a ref? ref seems better for all things that don't need to cause a render.

Reply:
yeah, i'm not sure why they implement it with state, I've moved it over anyways. they should have the same behavior